treq/testing.py

```diff
@@ -205,3 +207,218 @@ def __init__(self, resource):
+    in its place).  The callable, or whomever creates the callable, should
+    have a way to handle error reporting.
+    """
+    isLeaf = True
+
+    def __init__(self, get_response_for):
+        """
+        :param get_response_for: A callable that takes 5 parameters:
+        ``(method, url, params, headers, data) -> (code, headers, body)``
+        """
+        Resource.__init__(self)
+        self._get_response_for = get_response_for
+
+    def render(self, request):
+        """
+        Produce a response according to the stubs provided.
+        """
+        params = request.args
+        headers = {}
+        for k, v in request.requestHeaders.getAllRawHeaders():
+            headers[k] = v
+
+        for dictionary in (params, headers):
+            for k in dictionary:
+                dictionary[k] = sorted(dictionary[k])
+
+        # The incoming request does not have the absoluteURI property, because
+        # an incoming request is a IRequest, not an IClientRequest, so it
+        # the absolute URI needs to be synthesized.
+
+        # But request.URLPath() only returns the scheme and hostname, because
+        # that is the URL for this resource (because this resource handles
+        # everything from the root on down).
+
+        # So we need to add the request.path (not request.uri, which includes
+        # the query parameters)
+        absoluteURI = str(request.URLPath().click(request.path))
+
+        status_code, headers, body = self._get_response_for(
+            request.method, absoluteURI, params, headers,
+            request.content.read())
+
+        request.setResponseCode(status_code)
+        for k, v in headers.items():
+            request.setHeader(k, v)
+
+        return body
+
+
+class HasHeaders(object):
+    """
+    Since Twisted adds headers to a request, such as the host and the content
+    length, it's necessary to test whether request headers CONTAIN the expected
+    headers (the ones that are not automatically added by Twisted).
+
+    This wraps a set of headers, and can be used in an equality test against
+    a superset if the provided headers.
+    """
+    def __init__(self, headers):
+        self._headers = dict([(k.lower(), v) for k, v in headers.items()])
+
+    def __repr__(self):
+        return "HasHeaders({0})".format(repr(self._headers))
+
+    def __eq__(self, other_headers):
+        compare_to = dict([(k.lower(), v) for k, v in other_headers.items()])
+
+        return (set(self._headers.keys()).issubset(set(compare_to.keys())) and
+                all([set(v).issubset(set(compare_to[k]))
+                     for k, v in self._headers.items()]))
+
+    def __ne__(self, other_headers):
+        return not self.__eq__(other_headers)
+
+
+class RequestSequence(object):
+    """
+    Takes a sequence of::
+
+        [((method, url, params, headers, data), (code, headers, body)),
+         ...]
+
+    Expects the requests to arrive in sequence order.  If there are no more
```

Comment:
I think this should say something, early in the docstring, like "for an example of usage, see consume()", so that someone reading this doc can quickly figure out what this is *for* rather than just what it does.